File: gopath/src/bitwrk/client/worker.go
//  BitWrk - A Bitcoin-friendly, anonymous marketplace for computing power
//  Copyright (C) 2013  Jonas Eschenburg <jonas@bitwrk.net>
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

package client

import (
	"bitwrk"
	"log"
	"sync"
	"time"
)

type WorkerManager struct {
	mutex           sync.Mutex
	workers         map[string]*WorkerState
	activityManager *ActivityManager
	receiveManager  *ReceiveManager
}

type WorkerState struct {
	m            *WorkerManager
	Info         WorkerInfo
	stopOffering func()
}

type WorkerInfo struct {
	Id      string
	Article bitwrk.ArticleId
	Method  string
	PushURL string
}

func NewWorkerManager(a *ActivityManager, r *ReceiveManager) *WorkerManager {
	m := new(WorkerManager)
	m.workers = make(map[string]*WorkerState)
	m.activityManager = a
	m.receiveManager = r
	return m
}

func (m *WorkerManager) ListWorkers() (result []WorkerState) {
	m.mutex.Lock()
	defer m.mutex.Unlock()
	result = make([]WorkerState, 0, len(m.workers))
	for _, workerState := range m.workers {
		result = append(result, *workerState)
	}
	return
}

func (m *WorkerManager) RegisterWorker(info WorkerInfo) {
	m.mutex.Lock()
	defer m.mutex.Unlock()
	if s, ok := m.workers[info.Id]; ok {
		log.Printf("Changed worker %#v: %v -> %v", info.Id, s.Info, info)
		s.Info = info
	} else {
		log.Printf("Registered worker: %v", info)
		chStop := make(chan int, 1)
		s = &WorkerState{
			m:    m,
			Info: info,
			stopOffering: func() {
				// The value has no meaning
				chStop <- 0
			},
		}
		m.workers[info.Id] = s
		go s.keepOffering(chStop)
	}
}

func (m *WorkerManager) UnregisterWorker(id string) {
	m.mutex.Lock()
	s, ok := m.workers[id]
	delete(m.workers, id)
	m.mutex.Unlock()

	if ok {
		s.stopOffering()
	}
}

func (s *WorkerState) keepOffering(chStop <-chan int) {
	log.Printf("Start offering worker %#v", s.Info.Id)
	defer log.Printf("Stopped offering worker %#v", s.Info.Id)
	for {
		select {
		case _ = <-chStop:
			return
		default:
			s.offer()
		}
	}
}

func (s *WorkerState) offer() {
	log.Printf("Offering worker %#v", s.Info.Id)
	if sell, err := s.m.activityManager.NewSell(&s.Info); err != nil {
		log.Printf("Error creating sell: %v", err)
	} else {
		log.Printf("Performing sell")
		if err = sell.Perform(s.m.receiveManager); err != nil {
			log.Printf("Error performing sell: %v", err)
			time.Sleep(20 * time.Second)
		}
	}
}

File: bitwrk-blender/render_bitwrk/worker.py
# ##### BEGIN GPL LICENSE BLOCK #####
#
#  BitWrk - A Bitcoin-friendly, anonymous marketplace for computing power
#  Copyright (C) 2013-2018  Jonas Eschenburg <jonas@bitwrk.net>
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# ##### END GPL LICENSE BLOCK #####

import atexit, bpy.app, os.path, subprocess

WORKER_PROC = None
worker_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), "blender-slave.py")

def _exithandler():
    global WORKER_PROC
    p, WORKER_PROC = WORKER_PROC, None
    if p is not None:
        p.terminate()
        
def worker_alive():
    global WORKER_PROC
    if WORKER_PROC is None:
        return False
    if WORKER_PROC.poll() is not None:
        # process has exited
        WORKER_PROC = None
        return False
    return True

def can_start_worker(settings):
    if worker_alive():
        return False
    return os.path.isfile(worker_path)

def start_worker(settings):
    global WORKER_PROC
    if worker_alive():
        return
    if settings.use_custom_python_executable:
        python_path = settings.custom_python_executable
    else:
        python_path = bpy.app.binary_path_python 
    args = [
        python_path,
        worker_path,
        "--blender", bpy.app.binary_path,
        "--bitwrk-host", str(settings.bitwrk_client_host),
        "--bitwrk-port", str(settings.bitwrk_client_port),
        "--max-cost", str(settings.complexity),
        "--device", str(settings.worker_device),
        ]
    print("Starting worker:", args)
    WORKER_PROC = subprocess.Popen(args)
    atexit.register(_exithandler)
    
def can_stop_worker():
    return worker_alive()

def stop_worker():
    global WORKER_PROC
    if not worker_alive():
        return
    print("Terminating worker")
    p, WORKER_PROC = WORKER_PROC, None
    p.terminate()
    atexit.unregister(_exithandler)
    
